java/core/lance-jni/src/blocking_dataset.rs

```diff
@@ -146,6 +149,74 @@ pub extern "system" fn Java_com_lancedb_lance_Dataset_importFfiSchema(
     };
 }
 
+#[no_mangle]
+pub extern "system" fn Java_com_lancedb_lance_ipc_DatasetScanner_openStream(
+    mut env: JNIEnv,
+    _reader: JObject,
+    jdataset: JObject,
+    columns: JObject,       // Optional<String[]>
+    substrait_filter_obj: JObject, // Optional<ByteBuffer>
+    filter_obj: JObject,    // Optional<String>
+    batch_size: jlong,
+    stream_addr: jlong,
+) {
+    let columns = ok_or_throw_without_return!(env, env.get_strings_opt(&columns));
+    let dataset = {
+        let dataset =
+            unsafe { env.get_rust_field::<_, _, BlockingDataset>(jdataset, NATIVE_DATASET) }
+                .expect("Dataset handle not set");
+        dataset.clone()
+    };
+    let mut scanner: Scanner = dataset.inner.scan();
+    if let Some(cols) = columns {
+        ok_or_throw_without_return!(env, scanner.project(&cols));
+    };
+    let substrait = ok_or_throw_without_return!(env, env.get_bytes_opt(&substrait_filter_obj));
+    if let Some(substrait) = substrait {
+        ok_or_throw_without_return!(
+            env,
+            RT.block_on(async { scanner.filter_substrait(substrait).await })
+        );
+    }
+    let filter = ok_or_throw_without_return!(env, env.get_string_opt(&filter_obj));
+    if let Some(filter) = filter {
+        ok_or_throw_without_return!(env, scanner.filter(filter.as_str()));
+    }
+    scanner.batch_size(batch_size as usize);
+
+    let stream =
+        ok_or_throw_without_return!(env, RT.block_on(async { scanner.try_into_stream().await }));
+    let ffi_stream = to_ffi_arrow_array_stream(stream, RT.handle().clone()).unwrap();
+    unsafe { std::ptr::write_unaligned(stream_addr as *mut FFI_ArrowArrayStream, ffi_stream) }
+}
+
+#[no_mangle]
+pub extern "system" fn Java_com_lancedb_lance_ipc_DatasetScanner_getSchema(
+    mut env: JNIEnv,
+    _scanner: JObject,
+    jdataset: JObject,
+    columns: JObject, // Optional<String[]>
+) -> jlong {
+    let columns = ok_or_throw_with_return!(env, env.get_strings_opt(&columns), -1);
+
+    let res = {
+        let dataset =
+            unsafe { env.get_rust_field::<_, _, BlockingDataset>(jdataset, NATIVE_DATASET) }
+                .expect("Dataset handle not set");
+        dataset.clone()
+    };
+    let ds_schema = res.inner.schema();
+    let schema = if let Some(columns) = columns {
+        ok_or_throw_with_return!(env, ds_schema.project(&columns), -1)
+    } else {
+        ds_schema.clone()
+    };
+    let arrow_schema: arrow::datatypes::Schema = (&schema).into();
+    let ffi_schema =
+        Box::new(FFI_ArrowSchema::try_from(&arrow_schema).expect("Failed to convert schema"));
+    Box::into_raw(ffi_schema) as jlong
```

Comment:
does this leak?

i.e. when we call `ArrowSchema.wrap(addr)`, does the JVM arrow take care of freeing this memory?

Reply:
There are two approaches about getting FFI Schema in the JNI code base
**Approach 1: Java creates & manages ArrowSchema, pass to Rust to fill in**
Rust Side
```
#[no_mangle]
pub extern "system" fn Java_com_lancedb_lance_Dataset_importFfiSchema(
    mut env: JNIEnv,
    jdataset: JObject,
    arrow_schema_addr: jlong,
) {
    let schema = {
        let dataset =
            unsafe { env.get_rust_field::<_, _, BlockingDataset>(jdataset, NATIVE_DATASET) }
                .expect("Failed to get native dataset handle");
        Schema::from(dataset.inner.schema())
    };
    let out_c_schema = arrow_schema_addr as *mut FFI_ArrowSchema;
    # The top level s_chema is stack_allocator, will be cleanup by Rust when out of scope
    let c_schema = ok_or_throw_without_return!(env, FFI_ArrowSchema::try_from(&schema)); 

    unsafe {
        std::ptr::copy(std::ptr::addr_of!(c_schema), out_c_schema, 1);
        std::mem::forget(c_schema);  
    };
}
```
After copy
- out_c_schema address -> (points to) -> actual object
- c_schema address -> (points to) -> actual object
After forget
- actual object will not be cleaned by Rust deconstructor and will be handled by Java Side
- c_schema  address is stack allocated, will be cleanuped by Rust when out of scope

On the 
```
Java side
```
  public Schema getSchema() {
    try (ArrowSchema ffiArrowSchema = ArrowSchema.allocateNew(allocator)) {
      importFfiSchema(ffiArrowSchema.memoryAddress());
      return Data.importSchema(allocator, ffiArrowSchema, null);
    }
  }

  private native void importFfiSchema(long arrowSchemaMemoryAddress);
```
On the Java side, ArrowSchema full lifecycle is managed
- ffiArrowSchema will be allocated and cleanup through try with resources
- Data.importSchema will release the actual object of arrowSchema and close the given ArrowSchema (The only thing is we may be double close the ArrowSchema)

Reply:
Approach 2: Rust creates the ArrowSchema and pass to Java
```
pub extern "system" fn Java_com_lancedb_lance_ipc_DatasetScanner_getSchema(
) -> jlong {
    let arrow_schema: arrow::datatypes::Schema = (&schema).into();
    let ffi_schema =
        Box::new(FFI_ArrowSchema::try_from(&arrow_schema).expect("Failed to convert schema"));
    Box::into_raw(ffi_schema) as jlong
}
```
Java Side
```
    long address = getSchema(dataset, options.getColumns());
    try (ArrowSchema schema = ArrowSchema.wrap(address)) {
      return Data.importSchema(allocator, schema, null);
    }
```
This approach i am not quite sure whether it's memory safe with unknown behavior of `FFI_ArrowSchema::try_from`

Will consolidate and modify to approach 1
@eddyxu FYI

Reply:
Changed to approach 1 and consolidate part of the code

Reply:
yeah, I wonder if we need some sort of `Closeable` that calls rust when on `.close`.

Reply:
Follow the c interface convention of producer (java) create the arrow schema, release the arrow schema (via Data.exportSchema which calls schema.release() & schema.close)

On the Rust side, making sure we swap the Java empty ArrowSchema with the Rust created real-data ArrowSchema, so that Rust will fully manage the lifecycle of empty FFI_ArrowSchema 

Limited examples available for the Rust FFI_ArrowSchema handling example [1](https://sourcegraph.com/github.com/lakesoul-io/LakeSoul@a351fb4a284947e9ab3d0af3db5241fe47e24640/-/blob/rust/lakesoul-io-c/src/lib.rs?L628:39-628:62) [2](https://sourcegraph.com/github.com/qslang/queryscript@b48fcc173fbaa7edea7d2fb44f1332ea11240b2a/-/blob/queryscript/src/runtime/duckdb/engine.rs?L537:50-537:73) [3](https://sourcegraph.com/github.com/pacman82/arrow-odbc-py@ae233587abe0281356febfdaafa5bae788cea677/-/blob/src/reader.rs?L186:25-186:48) 

The release callback should be added directly to FFI_ArrowSchema when created it, didn't find too much document about the Rust FFI_ArrowSchema